### PR TITLE
Pass span context from runtime to agent to get a full trace #1968

### PR DESCRIPTION
**Pull request by liubin**
Pass span context from runtime to agent to get a full trace #1968

Now WIP, but review is welcome. @jodh-intel @cmaf

### CODE REVIEW

**Comment by jodh-intel** [comment]
@liubin - this sounds like [collated tracing](https://github.com/kata-containers/kata-containers/pull/1937). If so, can you add a fixes comment for https://github.com/kata-containers/kata-containers/issues/420?

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -51,6 +51,15 @@ use crate::sandbox::Sandbox;
 use crate::version::{AGENT_VERSION, API_VERSION};
 use crate::AGENT_CONFIG;
 
+// this ueses are for trace_rpc_call
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:54]: I'm not sure this comment adds much so I'd consider deleting it.

File: src/agent/src/tracer.rs [hunk 2]
```diff
@@ -81,6 +84,8 @@ pub fn setup_tracing(name: &'static str, logger: &Logger, _agent_cfg: &AgentConf
 
     tracing::subscriber::set_global_default(subscriber)?;
 
+    global::set_text_map_propagator(TraceContextPropagator::new());
```
Comment (jodh-intel) [inline src/agent/src/tracer.rs:87]: If we're still going to allow for ["isolated" agent tracing](https://github.com/kata-containers/kata-containers/pull/1937), I think this is where we need to check if `TraceType::Collated` has been requested.

See: https://github.com/kata-containers/kata-containers/pull/1937#issuecomment-855812254

**Review by jodh-intel — COMMENTED** [review]
@liubin - Nice!

Since I still haven't got [the shutdown test to land](https://github.com/kata-containers/kata-containers/pull/1809), could you test to ensure the agent can still be shutdown with this PR?

Also, does the Jaeger output look better with this PR? (I suspect yes ;)

**Comment by liubin** [comment]
> @liubin - Nice!
> 
> Since I still haven't got [the shutdown test to land](https://github.com/kata-containers/kata-containers/pull/1809), could you test to ensure the agent can still be shutdown with this PR?
> 
> Also, does the Jaeger output look better with this PR? (I suspect yes ;)

Yes, the shutdown has worked correctly.

And here is an image of one trace:

image

I'd advise testing it on local env to check the real traces in Jaeger if you have time.

**Comment by jodh-intel** [comment]
Hi @liubin - nice! I'm having trouble getting the same results though. Enabling runtime and agent tracing, I see two Jaeger services:

- `kata`:  which only shows `ttrpc.WriteStdin` spans.
- `kata-agent`: which only show `find_process()` and `get_container()` spans.

It's also surprising I'm seeing two services since I would have expectec to only see one (`kata`)? How have you set your config?

I'll try creating a fresh environment and re-testing this PR next week as I'm out tomorrow.

/cc @cmaf in case she has a chance to test this PR between now and then.

**Comment by liubin** [comment]
Hi @jodh-intel 

In runtime, the service name is set by this line:

https://github.com/kata-containers/kata-containers/blob/dc4307d3cc950bf552ef3e3a3cb148c9abc0032f/src/runtime/containerd-shim-v2/create.go#L72

And in the agent side, this is set in trace forwarder:

https://github.com/kata-containers/kata-containers/blob/dc4307d3cc950bf552ef3e3a3cb148c9abc0032f/src/trace-forwarder/src/tracer.rs#L28-L32

You have done a great job, that it can be configured through the `--trace-name`  option, though the name should be "service-name".

But IMO, using different service names for runtime/agent will be distinct to understand.

Below is my configuration.toml file for tracing:

```toml
[runtime]
enable_tracing = true
jaeger_endpoint="http://1.2.3.4:14268/api/traces"

[agent.kata]
enable_tracing = true
trace_mode = "static"
```

**Comment by cmaf** [comment]
> In runtime, the service name is set by this line:
> 
> https://github.com/kata-containers/kata-containers/blob/dc4307d3cc950bf552ef3e3a3cb148c9abc0032f/src/runtime/containerd-shim-v2/create.go#L72

@liubin @jodh-intel Maybe we should change this to "kata-runtime"? Or make it configurable like in the agent...

File: src/agent/src/sandbox.rs [hunk 3]
```diff
@@ -181,7 +181,6 @@ impl Sandbox {
         Ok(true)
     }
 
-    #[instrument]
     pub fn add_container(&mut self, c: LinuxContainer) {
```
Comment (cmaf) [inline src/agent/src/sandbox.rs:184]: Does instrumenting `add_container`, `get_container`, and `find_process` produce a lot of output? I agree that they are simple enough that we don't need to trace them, but I'm wondering if having a few isolated spans in other cases is really a problem.

**Review by cmaf — COMMENTED** [review]
@liubin Question about removing instrument attribute.

File: src/agent/src/sandbox.rs (hunk 3 above)
Comment (liubin) [reply src/agent/src/sandbox.rs:184]: Precisely to say is that `get_container` generates too many spans. RPC calls `write_stdin`/`read_stdout`/`read_stderr` now don't generate any spans, this calls all call `get_container`.

Isolated spans are not an important problem(or bug), they only are split from their call hierarchy.

**Review by jodh-intel — APPROVED** [review]
Thanks @liubin. I re-tested and now see a hierarchy of spans similar to your screenshot.

lgtm!

**Comment by liubin** [comment]
@jodh-intel @cmaf I have two additional commits [here](https://github.com/liubin/kata-containers/pull/8) to resolve some minor issues, can you take a look to check if these commits should be included in this PR.

**Comment by jodh-intel** [comment]
Thanks @liubin. I've commented on https://github.com/liubin/kata-containers/pull/8. We could always just land this PR anyway and then add the 2nd one later potentially?

**Comment by jodh-intel** [comment]
Ping for reviews y'all! ;)

File: src/agent/src/tracer.rs (hunk 2 above)
Comment (fidencio) [reply src/agent/src/tracer.rs:87]: From the discussion we had during the Architecture Committee meetings seems that "isolated" is not the goal, at least not in the first interactions of this, right?

File: src/runtime/virtcontainers/kata_agent.go [hunk 4]
```diff
@@ -2014,15 +2029,15 @@ func (k *kataAgent) installReqFunc(c *kataclient.AgentClient) {
 	}
 }
 
-func (k *kataAgent) getReqContext(reqName string) (ctx context.Context, cancel context.CancelFunc) {
-	ctx = context.Background()
+func (k *kataAgent) getReqContext(ctx1 context.Context, reqName string) (ctx context.Context, cancel context.CancelFunc) {
```
Comment (fidencio) [inline src/runtime/virtcontainers/kata_agent.go:2032]: Please, use "ctx" and "newCtx" instead of "ctx1" and "ctx"

**Review by fidencio — APPROVED** [review]
lgtm, but feel free to take it with a grain of salt, due to my limited knowledge on the area.

File: src/runtime/virtcontainers/kata_agent.go (hunk 4 above)
Comment (liubin) [reply src/runtime/virtcontainers/kata_agent.go]: @fidencio have updated as your advice.

**Comment by jodh-intel** [comment]
\o/ ;)

**Pull request merged**